File: platform/challenges/svc3-specterlog/service/src/ratelimit.ts
import type { Database } from "bun:sqlite";
import type { FastifyInstance, FastifyReply, FastifyRequest } from "fastify";

export type Bucket = {
  capacity: number;
  refillPerSecond: number;
};

const DEFAULTS: Record<string, Bucket> = {
  default: { capacity: 60, refillPerSecond: 1.0 },
  auth: { capacity: 10, refillPerSecond: 0.2 },
  upload: { capacity: 12, refillPerSecond: 0.5 },
};

function clientKey(req: FastifyRequest, route: string): string {
  const forwarded = req.headers["x-forwarded-for"];
  const ip = Array.isArray(forwarded)
    ? forwarded[0]
    : (typeof forwarded === "string" ? forwarded.split(",")[0].trim() : null)
      ?? (req.ip ?? "unknown");
  return `${ip}|${route}`;
}

export function consume(
  db: Database,
  key: string,
  bucket: Bucket,
): { ok: boolean; remaining: number; retry_after: number } {
  const now = Date.now();
  const row = db.query(
    "SELECT tokens, refilled_at FROM rate_buckets WHERE key = ?",
  ).get(key) as { tokens: number; refilled_at: number } | null;
  let tokens = row ? row.tokens : bucket.capacity;
  const last = row ? row.refilled_at : now;
  const elapsed = Math.max(0, now - last) / 1000;
  tokens = Math.min(bucket.capacity, tokens + elapsed * bucket.refillPerSecond);
  if (tokens < 1) {
    db.run(
      "INSERT OR REPLACE INTO rate_buckets (key, tokens, refilled_at) VALUES (?, ?, ?)",
      [key, tokens, now],
    );
    const deficit = 1 - tokens;
    const retry = bucket.refillPerSecond > 0 ? Math.ceil(deficit / bucket.refillPerSecond) : 60;
    return { ok: false, remaining: 0, retry_after: retry };
  }
  tokens -= 1;
  db.run(
    "INSERT OR REPLACE INTO rate_buckets (key, tokens, refilled_at) VALUES (?, ?, ?)",
    [key, tokens, now],
  );
  return { ok: true, remaining: Math.floor(tokens), retry_after: 0 };
}

export function enforce(
  db: Database,
  req: FastifyRequest,
  reply: FastifyReply,
  route: string,
  bucketName: string = "default",
): boolean {
  const bucket = DEFAULTS[bucketName] ?? DEFAULTS.default;
  const result = consume(db, clientKey(req, route), bucket);
  if (!result.ok) {
    reply.code(429).header("retry-after", String(result.retry_after)).send({
      error: "rate_limited",
      retry_after: result.retry_after,
    });
    return false;
  }
  reply.header("x-ratelimit-remaining", String(result.remaining));
  return true;
}

export function rateRoutes(app: FastifyInstance, db: Database): void {
  app.get("/api/rate-limit/status", async (req) => {
    const key = clientKey(req, "status");
    const row = db.query(
      "SELECT tokens, refilled_at FROM rate_buckets WHERE key = ?",
    ).get(key) as { tokens: number; refilled_at: number } | null;
    const bucket = DEFAULTS.default;
    const now = Date.now();
    const tokens = row
      ? Math.min(bucket.capacity, row.tokens + Math.max(0, now - row.refilled_at) / 1000 * bucket.refillPerSecond)
      : bucket.capacity;
    return {
      bucket: "default",
      capacity: bucket.capacity,
      remaining: Math.floor(tokens),
      refill_per_second: bucket.refillPerSecond,
    };
  });

  app.addHook("onRequest", async (req, reply) => {
    const url = req.url ?? "";
    if (url.startsWith("/api/accounts/register") || url.startsWith("/api/accounts/login")) {
      if (!enforce(db, req, reply, "auth", "auth")) return;
    } else if (url.includes("/attachments") && req.method === "POST") {
      if (!enforce(db, req, reply, "upload", "upload")) return;
    } else if (url === "/" && req.method === "POST") {
      // checker RPC; auth is the secret header
      return;
    } else if (url.startsWith("/api/")) {
      if (!enforce(db, req, reply, "default", "default")) return;
    }
  });
}

File: platform/challenges/svc1-ledgerforge/service/src/ratelimit.rs
use std::sync::Arc;

use axum::{
    extract::{Request, State},
    http::StatusCode,
    middleware::Next,
    response::{IntoResponse, Response},
    Json,
};
use rusqlite::params;
use serde_json::json;
use tokio::sync::Mutex;

use crate::{db, state::AppState};

pub type Shared = Arc<Mutex<AppState>>;

#[derive(Clone, Copy)]
struct Bucket {
    capacity: f64,
    refill_per_second: f64,
}

const DEFAULT: Bucket = Bucket { capacity: 60.0, refill_per_second: 1.0 };
const AUTH: Bucket = Bucket { capacity: 10.0, refill_per_second: 0.2 };
const WRITE: Bucket = Bucket { capacity: 12.0, refill_per_second: 0.5 };

fn pick(uri_path: &str, method: &str) -> Option<(Bucket, &'static str)> {
    if uri_path.starts_with("/api/accounts/register") || uri_path.starts_with("/api/accounts/login") {
        return Some((AUTH, "auth"));
    }
    if uri_path == "/" && method == "POST" {
        return None;
    }
    if uri_path.starts_with("/api/") && matches!(method, "POST" | "PATCH" | "PUT" | "DELETE") {
        return Some((WRITE, "write"));
    }
    None
}

fn consume(state: &AppState, key: &str, bucket: Bucket) -> Option<i64> {
    let now = db::now_ms();
    let row: rusqlite::Result<(f64, i64)> = state.db.query_row(
        "SELECT tokens, refilled_at FROM rate_buckets WHERE key = ?1",
        params![key],
        |r| Ok((r.get(0)?, r.get(1)?)),
    );
    let (mut tokens, last) = row.unwrap_or((bucket.capacity, now));
    let elapsed_ms = (now - last).max(0) as f64;
    tokens = (tokens + (elapsed_ms / 1000.0) * bucket.refill_per_second).min(bucket.capacity);
    if tokens < 1.0 {
        let _ = state.db.execute(
            "INSERT OR REPLACE INTO rate_buckets (key, tokens, refilled_at) VALUES (?1, ?2, ?3)",
            params![key, tokens, now],
        );
        let deficit = 1.0 - tokens;
        let retry = if bucket.refill_per_second > 0.0 {
            (deficit / bucket.refill_per_second).ceil() as i64
        } else {
            60
        };
        return Some(retry);
    }
    tokens -= 1.0;
    let _ = state.db.execute(
        "INSERT OR REPLACE INTO rate_buckets (key, tokens, refilled_at) VALUES (?1, ?2, ?3)",
        params![key, tokens, now],
    );
    None
}

pub async fn middleware(State(state): State<Shared>, request: Request, next: Next) -> Response {
    let uri_path = request.uri().path().to_string();
    let method = request.method().as_str().to_string();
    if let Some((bucket, class)) = pick(&uri_path, &method) {
        let ip = request
            .headers()
            .get("x-forwarded-for")
            .and_then(|v| v.to_str().ok())
            .and_then(|v| v.split(',').next())
            .map(|s| s.trim().to_string())
            .unwrap_or_else(|| "unknown".to_string());
        let key = format!("{ip}|{class}");
        let guard = state.lock().await;
        if let Some(retry) = consume(&guard, &key, bucket) {
            drop(guard);
            return (
                StatusCode::TOO_MANY_REQUESTS,
                [("retry-after", retry.to_string())],
                Json(json!({"error": "rate_limited", "retry_after": retry})),
            )
                .into_response();
        }
    }
    next.run(request).await
}

pub async fn status(State(state): State<Shared>) -> impl IntoResponse {
    let state = state.lock().await;
    Json(json!({"bucket": "default", "capacity": DEFAULT.capacity, "refill_per_second": DEFAULT.refill_per_second, "ts": db::now_ms()}))
}
